hypershift-operator/controllers/nodepool/inplace.go

```diff
@@ -0,0 +1,402 @@
+	"sigs.k8s.io/controller-runtime/pkg/client/apiutil"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+	"sigs.k8s.io/controller-runtime/pkg/source"
+)
+
+// reconcileInPlaceUpgrade loops over all Nodes that belong to a NodePool and performs an in place upgrade if necessary.
+func (r *NodePoolReconciler) reconcileInPlaceUpgrade(ctx context.Context, hc *hyperv1.HostedCluster, nodePool *hyperv1.NodePool, machineSet *capiv1.MachineSet, targetConfigHash, targetVersion, targetConfigVersionHash string) error {
+	log := ctrl.LoggerFrom(ctx)
+
+	// If there's no guest cluster yet return early.
+	if hc.Status.KubeConfig == nil {
+		return nil
+	}
+
+	hostedClusterClient, err := newHostedClusterClient(ctx, r.Client, hc)
+	if err != nil {
+		return fmt.Errorf("failed to create remote client: %v", err)
+	}
+
+	// Watch hosted cluster Nodes. We track the created caches, so we don't add a watcher on every reconciliation.
+	r.hostedClusterCachesTracker.Lock()
+	defer r.hostedClusterCachesTracker.Unlock()
+	if !r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(nodePool)] {
+		hostedClusterCache, err := newHostedClusterCache(ctx, r.Client, hc)
+		if err != nil {
+			return fmt.Errorf("failed to create hosted cluster cache: %v", err)
+		}
+
+		// TODO (alberto): cancel the ctx on exit.
+		go hostedClusterCache.Start(ctx)
+		if !hostedClusterCache.WaitForCacheSync(ctx) {
+			return fmt.Errorf("failed waiting for hosted cluster cache to sync: %w", err)
+		}
+
+		if err := r.controller.Watch(source.NewKindWithCache(&corev1.Node{}, hostedClusterCache), handler.EnqueueRequestsFromMapFunc(r.nodeToNodePool)); err != nil {
+			return fmt.Errorf("error adding watcher for hosted cluster nodes: %v", err)
+		}
+
+		// TODO: index by HC here instead?
+		if r.hostedClusterCachesTracker.caches == nil {
+			r.hostedClusterCachesTracker.caches = make(map[client.ObjectKey]bool)
+		}
+		r.hostedClusterCachesTracker.caches[client.ObjectKeyFromObject(nodePool)] = true
+		log.Info("Created hosted cluster cache")
+	}
+
+	nodes, err := getNodesForMachineSet(ctx, r.Client, hostedClusterClient, machineSet)
+	if err != nil {
+		return err
+	}
+
+	// If all Nodes are atVersion
+	if inPlaceUpgradeComplete(nodes, targetConfigVersionHash) {
+		if nodePool.Status.Version != targetVersion {
+			log.Info("Version update complete",
+				"previous", nodePool.Status.Version, "new", targetVersion)
+			nodePool.Status.Version = targetVersion
+		}
+
+		if nodePool.Annotations[nodePoolAnnotationCurrentConfig] != targetConfigHash {
+			log.Info("Config update complete",
+				"previous", nodePool.Annotations[nodePoolAnnotationCurrentConfig], "new", targetConfigHash)
+			nodePool.Annotations[nodePoolAnnotationCurrentConfig] = targetConfigHash
+		}
+		nodePool.Annotations[nodePoolAnnotationCurrentConfigVersion] = targetConfigVersionHash
+		return nil
+	}
+
+	// Otherwise:
+	// Order Nodes deterministically.
```

Comment:
what happens if new nodes are added while this is occurring?

Reply:
So this is a behaviour difference between MCO today and Hypershift. In Hypershift, there is no "previous config". The ignition server will always sync first. Once the switch occurs, the nodepool will scale immediately into the new config.

The MCO today instead waits until at least 1 node finishes the update, before serving the new config. So during that inbetween, the nodes scale up into the new config.

From the MCD perspective this shouldn't matter. If we are already on the version we want, we will simply not update. So new nodes joining will eventually attempt to get updated here, and be a no-op (other than indicate success)

Side note: I'm not sure what ordering nodes deterministically entails. The controller should just be finding un-upgraded nodes and upgrading them. The ordering shouldn't really matter.

Reply:
New nodes will come up with the new token secret payload, otherwise if a node is coming with outdated version inPlaceUpgradeComplete should return false.

Reply:
>Side note: I'm not sure what ordering nodes deterministically entails. The controller should just be finding un-upgraded nodes and upgrading them. The ordering shouldn't really matter.

We'll need to honour maxUnavailable/maxSurge, for that we'll need to know when a Node upgrade is complete or return early and reconcile, for that deterministic ordering comes in handy. In any case that's up to the follow up implementation, let's discuss in that PR.